doc/man/fedfs-create-junction.8
.\"@(#)fedfs-create-junction.8"
.\"
.\" @file doc/man/fedfs-create-junction.8
.\" @brief man page for fedfs-create-junction client command
.\"

.\"
.\" Copyright 2011 Oracle.  All rights reserved.
.\"
.\" This file is part of fedfs-utils.
.\"
.\" fedfs-utils is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License version 2.0 as
.\" published by the Free Software Foundation.
.\"
.\" fedfs-utils is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License version 2.0 for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" version 2.0 along with fedfs-utils.  If not, see:
.\"
.\"	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
.\"
.TH FEDFS-CREATE-JUNCTION 8 "@publication-date@"
.SH NAME
fedfs-create-junction \- send a FEDFS_CREATE_JUNCTION ADMIN protocol request
.SH SYNOPSIS
.B fedfs-create-junction
.RB [ \-?d ]
.RB [ \-n
.IR nettype ]
.RB [ \-h
.IR hostname ]
.RB [ \-l
.IR nsdbname ]
.RB [ \-r
.IR nsdbport ]
.I path
.I fsn-uuid
.SH INTRODUCTION
RFC 5716 introduces the Federated File System (FedFS, for short).
FedFS is an extensible standardized mechanism
by which system administrators construct
a coherent namespace across multiple file servers using
.IR "file system referrals" .
For further details, see
.BR fedfs (7).
.P
FedFS-enabled file servers allow remote administrative access via an
authenticated RPC protocol known as the
.IR "FedFS ADMIN protocol" .
Using this protocol, FedFS administrators manage
FedFS junctions and NSDB connection parameter information
on remote FedFS-enabled file servers.
.SH DESCRIPTION
The
.BR fedfs-create-junction (8)
command is part of a collection of low-level single-use programs
that is intended for testing the FedFS ADMIN protocol or for use in scripts.
It sends a single FEDFS_CREATE_JUNCTION request to a remote
FedFS ADMIN protocol service.
.P
The FEDFS_CREATE_JUNCTION request creates a FedFS junction
in a local file system on a remote file server.
The contents of a FedFS junction are an FSN UUID and an NSDB name and port.
.P
The
.BR fedfs-create-junction (8)
command takes two positional parameters which specify
the pathname on the remote server of the new junction, and the FSN UUID.
This pathname is relative to the root
of the local file system on the remote server.
Required NSDB information can be inferred
from the command's environment or specified on the command line.
The meaning of these arguments is described in more detail in
.BR fedfs (7).
.P
The FEDFS_CREATE_JUNCTION request does not create an FSN record.
To create an FSN record, use the
.BR nsdb-create-fsn (8)
command.
Resolving a junction that contains an FSN UUID without
a matching FSN record on the NSDB results in an error on the file server.
.SH OPTIONS
.IP "\fB\-d, \-\-debug"
Enables debugging messages during operation.
.IP "\fB\-?, \-\-help"
Displays
.BR fedfs-create-junction (8)
version information and a usage message on
.IR stderr .
.IP "\fB\-h, \-\-hostname=\fIhostname\fP"
Specifies the hostname of a remote FedFS ADMIN service.
If this option is not specified, the default value is
.BR localhost .
.IP "\fB\-n, \-\-nettype=\fInettype\fP"
Specifies the transport to use when contacting the remote FedFS ADMIN service.
Typically the
.I nettype
is one of
.B tcp
or
.BR udp .
If this option is not specified, the default value is
.BR netpath .
See
.BR rpc (3t)
for details.
.IP "\fB\-l, \-\-nsdbname=\fINSDB-hostname\fP"
Specifies the hostname of the NSDB to insert into the new FedFS junction.
If this option is not specified,
the value of the FEDFS_NSDB_HOST environment variable is consulted.
If the variable is not set and the
.B \--nsdbname
option is not specified, the
.BR fedfs-create-junction (8)
command fails.
.IP "\fB\-r, \-\-nsdbport=\fINSDB-port\fP"
Specifies the IP port of the NSDB to insert into the new FedFS junction.
If this option is not specified,
the value of the FEDFS_NSDB_PORT environment variable is consulted.
The default value if the variable is not set is 389.
.SH EXIT CODES
The remote administration service returns a value that reflects the
success of the requested operation.
.TP
.B FEDFS_OK
The operation succeeded.
.TP
.B FEDFS_ERR_ACCESS
The caller does not have permission to perform the requested operation
on the remote server.
.TP
.B FEDFS_ERR_BADCHAR
The pathname contains a character which is not
supported by the remote server.
.TP
.B FEDFS_ERR_BADNAME
A component of the pathname consisted of valid UTF-8 characters
supported by the remote server,
but the name is not supported by the remote server
as a valid name for the current operation.
.TP
.B FEDFS_ERR_NAMETOOLONG
The length of the pathname exceeds the remote server’s implementation limit.
.TP
.B FEDFS_ERR_LOOP
Too many symbolic links were encountered in resolving the pathname.
.TP
.B FEDFS_ERR_BADXDR
The remote server encountered an XDR decoding error while
processing the request.
.TP
.B FEDFS_ERR_EXIST
The specified junction already exists.
.TP
.B FEDFS_ERR_INVAL
One of the arguments was not valid.
.TP
.B FEDFS_ERR_IO
A hard error occurred on the remote server.
.TP
.B FEDFS_ERR_NOSPC
The requested operation would have caused the remote
server’s filesystem to exceed some limit (for example, if there is
a fixed number of junctions per fileset or per server).
.TP
.B FEDFS_ERR_NOTJUNCT
The pathname does not end in a junction.
.TP
.B FEDFS_ERR_NOTLOCAL
The pathname contains a junction in any position other than the last component.
.TP
.B FEDFS_ERR_PERM
The operation was not allowed because the caller is
either not a privileged user or not the owner of an object that
would be modified by the operation.
.TP
.B FEDFS_ERR_ROFS
A modifying operation was attempted on a read-only filesystem.
.TP
.B FEDFS_ERR_SVRFAULT
An unanticipated non-protocol error occurred on the remote server.
.TP
.B FEDFS_ERR_NSDB_PARAMS
The remote server does not have any connection
parameters on record for the specified NSDB.
.TP
.B FEDFS_ERR_PATH_TYPE_UNSUPP
The remote server does not support the specified FedFsPathType value.
.TP
.B FEDFS_ERR_NOTSUPP
The remote server does not support the specified procedure.
.SH EXAMPLES
Suppose you are the FedFS administrator of the
.I example.net
FedFS domain and that your domain's NSDB hostname is
.IR nsdb.example.net .
To create a new FedFS junction on the file server
.IR fs.example.net ,
use:
.RS
.sp
$ fedfs-create-junction -h fs.example.net -l nsdb.example.net \\
	/export/junction1 `uuidgen -t`
.sp
.RE
In this example, a new FSN UUID is created on the spot.
It can be read back from the remote server using the
.BR fedfs-lookup-junction (8)
command, and added to the NSDB using the
.BR nsdb-create-fsn (8)
command.
.SH SECURITY
RPCSEC GSSAPI authentication has not yet been implemented for this command.
.SH "SEE ALSO"
.BR fedfs (7),
.BR rpc.fedfsd (8),
.BR fedfs-lookup-junction (8),
.BR nsdb-create-fsn (8),
.BR rpc (3t)
.sp
RFC 5716 for FedFS requirements and overview
.SH COLOPHON
This page is part of the fedfs-utils package.
A description of the project and information about reporting bugs
can be found at
.IR http://wiki.linux-nfs.org/wiki/index.php/FedFsUtilsProject .
.SH "AUTHOR"
Chuck Lever <chuck.lever@oracle.com>


src/fedfsc/fedfs-create-junction.c
/**
 * @file src/fedfsc/fedfs-create-junction.c
 * @brief Send a FEDFS_CREATE_JUNCTION RPC to a FedFS ADMIN server
 *
 * @todo
 *	Implement RPCGSS authentication
 */

/*
 * Copyright 2010 Oracle.  All rights reserved.
 *
 * This file is part of fedfs-utils.
 *
 * fedfs-utils is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2.0 as
 * published by the Free Software Foundation.
 *
 * fedfs-utils is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License version 2.0 for more details.
 *
 * You should have received a copy of the GNU General Public License
 * version 2.0 along with fedfs-utils.  If not, see:
 *
 *	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
 */

#include <sys/types.h>
#include <sys/stat.h>

#include <stdbool.h>
#include <fcntl.h>
#include <unistd.h>
#include <errno.h>
#include <getopt.h>
#include <locale.h>

#include <rpc/clnt.h>
#include <uuid/uuid.h>

#include "fedfs.h"
#include "fedfs_admin.h"
#include "nsdb.h"
#include "junction.h"
#include "xlog.h"
#include "gpl-boiler.h"

/**
 * Default RPC request timeout
 */
static struct timeval fedfs_create_junction_timeout = { 25, 0 };

/**
 * Short form command line options
 */
static const char fedfs_create_junction_opts[] = "?dh:l:n:r:";

/**
 * Long form command line options
 */
static const struct option fedfs_create_junction_longopts[] = {
	{ "debug", 0, NULL, 'd', },
	{ "help", 0, NULL, '?', },
	{ "hostname", 1, NULL, 'h', },
	{ "nettype", 1, NULL, 'n', },
	{ "nsdbname", 1, NULL, 'l', },
	{ "nsdbport", 1, NULL, 'r', },
	{ NULL, 0, NULL, 0, },
};

/**
 * Display program synopsis
 *
 * @param progname NUL-terminated C string containing name of program
 */
static void
fedfs_create_junction_usage(const char *progname)
{
	fprintf(stderr, "\n%s version " VERSION "\n", progname);
	fprintf(stderr, "Usage: %s [-d] [-n nettype] [-h hostname] "
			"[-l nsdbname] [-r nsdbport] "
			"pathname fsn-uuid\n\n",
			progname);

	fprintf(stderr, "\t-?, --help           Print this help\n");
	fprintf(stderr, "\t-d, --debug          Enable debug messages\n");
	fprintf(stderr, "\t-n, --nettype        RPC transport (default: 'netpath')\n");
	fprintf(stderr, "\t-h, --hostname       ADMIN server hostname (default: 'localhost')\n");
	fprintf(stderr, "\t-l, --nsdbname       NSDB hostname to set\n");
	fprintf(stderr, "\t-r, --nsdbport       NSDB port to set\n");

	fprintf(stderr, "%s", fedfs_gpl_boilerplate);

	exit((int)FEDFS_ERR_INVAL);
}

/**
 * Create a junction on a remote server
 *
 * @param hostname NUL-terminated UTF-8 string containing ADMIN server's hostname
 * @param nettype NUL-terminated C string containing nettype to use for connection
 * @param path NUL-terminated C string containing remote pathname of new junction
 * @param uuid NUL-terminated C string containing FSN UUID for new junction
 * @param nsdbname NUL-terminated UTF-8 string containing name of NSDB service for this junction
 * @param nsdbport port number of NSDB service for this junction
 * @return an exit status value
 */
static FedFsStatus
fedfs_create_junction_call(const char *hostname, const char *nettype,
		const char *path, const char *uuid, char *nsdbname,
		const uint16_t nsdbport)
{
	enum clnt_stat status;
	FedFsCreateArgs arg;
	FedFsStatus result;
	char **path_array;
	CLIENT *client;
	uuid_t uu;
	int res;

	memset(&arg, 0, sizeof(arg));

	res = uuid_parse(uuid, uu);
	if (res != 0) {
		fprintf(stderr, "Failed to parse UUID %s\n", uuid);
		return FEDFS_ERR_INVAL;
	}
	memcpy(arg.fsn.fsnUuid, uu, sizeof(FedFsUuid));

	arg.fsn.nsdbName.hostname.utf8string_val = nsdbname;
	arg.fsn.nsdbName.hostname.utf8string_len = strlen(nsdbname);
	arg.fsn.nsdbName.port = nsdbport;

	arg.path.type = FEDFS_PATH_SYS;
	result = nsdb_posix_to_path_array(path, &path_array);
	if (result != FEDFS_OK) {
		fprintf(stderr, "Failed to encode pathname: %s",
			nsdb_display_fedfsstatus(result));
		return result;
	}
	result = nsdb_path_array_to_fedfspathname(path_array,
				&arg.path.FedFsPath_u.adminPath);
	if (result != FEDFS_OK) {
		fprintf(stderr, "Failed to encode pathname: %s",
			nsdb_display_fedfsstatus(result));
		nsdb_free_string_array(path_array);
		return result;
	}

	client = clnt_create(hostname, FEDFS_PROG, FEDFS_V1, nettype);
	if (client == NULL) {
		clnt_pcreateerror("Failed to create FEDFS client");
		result = FEDFS_ERR_SVRFAULT;
		goto out;
	}

	memset((char *)&result, 0, sizeof(result));
	status = clnt_call(client, FEDFS_CREATE_JUNCTION,
				(xdrproc_t)xdr_FedFsCreateArgs,
				(caddr_t)&arg,
				(xdrproc_t)xdr_FedFsStatus, (caddr_t)&result,
				fedfs_create_junction_timeout);
	if (status != RPC_SUCCESS) {
		clnt_perror(client, "FEDFS_CREATE_JUNCTION call failed");
		result = FEDFS_ERR_SVRFAULT;
	} else
		nsdb_print_fedfsstatus(result);
	(void)clnt_destroy(client);

out:
	nsdb_free_fedfspathname(&arg.path.FedFsPath_u.adminPath);
	nsdb_free_string_array(path_array);
	return result;
}

/**
 * Program entry point
 *
 * @param argc count of command line arguments
 * @param argv array of NUL-terminated C strings containing command line arguments
 * @return program exit status
 */
int
main(int argc, char **argv)
{
	char *progname, *hostname, *nettype;
	char *fsn_uuid, *path, *nsdbname;
	unsigned short nsdbport;
	unsigned int seconds;
	FedFsStatus status;
	int arg;

	(void)setlocale(LC_ALL, "");
	(void)umask(S_IRWXO);

	/* Set the basename */
	if ((progname = strrchr(argv[0], '/')) != NULL)
		progname++;
	else
		progname = argv[0];

	/* For the libraries */
	xlog_stderr(1);
	xlog_syslog(0);
	xlog_open(progname);

	nsdb_env(&nsdbname, &nsdbport, NULL, NULL, NULL);

	hostname = "localhost";
	nettype = "netpath";
	while ((arg = getopt_long(argc, argv, fedfs_create_junction_opts,
			fedfs_create_junction_longopts, NULL)) != -1) {
		switch (arg) {
		case 'd':
			xlog_config(D_ALL, 1);
			break;
		case 'h':
			hostname = optarg;
			break;
		case 'l':
			if (!nsdb_is_hostname_utf8(optarg)) {
				fprintf(stderr, "NSDB name %s is "
					"not a UTF-8 hostname\n", optarg);
				fedfs_create_junction_usage(progname);
			}
			nsdbname = optarg;
			break;
		case 'n':
			nettype = optarg;
			break;
		case 'r':
			if (!nsdb_parse_port_string(optarg, &nsdbport)) {
				fprintf(stderr, "Bad port number: '%s'\n",
					optarg);
				fedfs_create_junction_usage(progname);
			}
			break;
		default:
			fprintf(stderr, "Invalid command line "
				"argument: %c\n", (char)arg);
		case '?':
			fedfs_create_junction_usage(progname);
		}
	}
	if (argc == optind + 2) {
		path = argv[optind];
		fsn_uuid = argv[optind + 1];
	} else {
		fprintf(stderr, "Ambiguous positional parameters\n");
		fedfs_create_junction_usage(progname);
	}
	if (nsdbname == NULL) {
		fprintf(stderr, "No NSDB hostname was specified\n");
		fedfs_create_junction_usage(progname);
	}

	for (seconds = FEDFS_DELAY_MIN_SECS;; seconds = fedfs_delay(seconds)) {
		status = fedfs_create_junction_call(hostname, nettype, path,
						fsn_uuid, nsdbname, nsdbport);
		if (status != FEDFS_ERR_DELAY)
			break;

		xlog(D_GENERAL, "Delaying %u seconds...", seconds);
		if (sleep(seconds) != 0)
			break;
	}
	return (int)status;
}
